### PR TITLE
Implement Account.onLogin/onLoginFailure on client.

### PR DESCRIPTION
**Pull request by jakozaur**
The change adds hooks `Account.onLogin`, `Accounts.onLoginFailure` on client. There are similar to their server equivalents.

Detailed discussion and rationale:
https://github.com/meteor/meteor/issues/3572
Tl;dr: Add an obvious way to implement client post login actions (such as `Router.go`).

Any comments are welcomed.

This is my first PR to meteor. Followed contributing guidelines and signed Meteor Contribution Agreement.

Testing:
- added two tests
- all test from this modules are passing `./meteor test-packages packages/accounts-base`
`./meteor test-packages packages/accounts-password`

### CODE REVIEW

**Comment by stubailo** [comment]
Hi, can you please construct a pull request with just the commit you intend to merge? (this one: https://github.com/jakozaur/meteor/commit/c9ea67ab836ef30edbe898fa5014ab83502906dd)

File: docs/client/full-api/api/accounts.md [hunk 1]
```diff
@@ -318,7 +318,7 @@ are called with a single argument, the attempt info object:
 
 <dl class="objdesc">
 {{#dtdd name="type" type="String"}}
-  The service name, such as "password" or "twitter".
+  (Server only) The service name, such as "password" or "twitter".
```
Comment (stubailo) [inline docs/client/full-api/api/accounts.md]: I'm not sure what these changes are - this is the documentation for `Accounts.validateLoginAttempt`, which isn't related to this PR?

Comment (stubailo) [inline docs/client/full-api/api/accounts.md]: Let's be more explicit for now, and I'll clean up the docs later (I got really confused about how this is related to this PR)

Something like:

> The service name, such as "password" or "twitter". For `Accounts.onLogin` and `Accounts.onLoginFailure`, this data is only passed on the server.

File: packages/accounts-base/accounts_common.js [hunk 2]
```diff
@@ -184,3 +184,30 @@ Accounts._tokenExpiresSoon = function (when) {
...
+  debugPrintExceptions: "onLogin callback"
+});
+onLoginFailureHook = new Hook({
+  debugPrintExceptions: "onLoginFailure callback"
+});
+
+
+/**
+ * @summary Register a callback to be called after a login attempt succeeds.
+ * @locus Anywhere
```
Comment (stubailo) [inline packages/accounts-base/accounts_common.js]: To get this change to stick, run `./scripts/admin/jsdoc/jsdoc.sh` from your Meteor checkout, then commit the new `data.js` file.

**Comment by jakozaur** [comment]
Thanks for comments @stubailo.

Implement the review suggestions and rerun the tests.

**Comment by stubailo** [comment]
@n1mmy can we confirm somehow that this code doesn't have any security implications?

**Comment by n1mmy** [comment]
Yes, this is code in a security-critical section. It should get extra review.

I've reviewed this and it looks good to me. The functional change is client-only, and thus shouldn't have any security impact.

Please get at least one other person to review and confirm there are no security implications. Also, please make sure @debergalis is at least aware of the change, though I don't expect he'll have particular opinions.

**Comment by stubailo** [comment]
@justinsb could you take a look at this? A fresh pair of eyes will be helpful to see if @n1mmy and I are missing something.

File: packages/accounts-base/accounts_client.js [hunk 3]
```diff
@@ -86,8 +86,29 @@ Accounts.callLoginMethod = function (options) {
...
   });
-  // make sure we only call the user's callback once.
-  var onceUserCallback = _.once(options.userCallback);
+  // Prepare callbacks: user provided and onLogin/onLoginFailure hooks.
+  var loginCallbacks = _.once(function (error) {
+    if (!error) {
+      onLoginHook.each(function (callback) {
+        callback({
+          allowed: true,
+          error: undefined,
```
Comment (glasser) [inline packages/accounts-base/accounts_client.js]: This is probably better as `null`.  See [the style guide](https://github.com/meteor/meteor/wiki/Meteor-Style-Guide#avoid-setting-variables-to-undefined----instead-use-null).

**Comment by justinsb** [comment]
I note that the callback now gets the method arguments (i.e. user password).  That makes be a bit nervous; I would rather be more defensive here and avoid passing it around too widely.

I understand that this makes the API the same on the client & on the server, but is there any other reason for including the password?  It may be that there is a big upside to passing the password around, which I am unaware of.  (And I think the correct order of proceedings here is to figure out whether there is a legitimate reason to include the password, before we even think about whether & how to exclude it!)

**Comment by stubailo** [comment]
> I note that the callback now gets the method arguments (i.e. user password). That makes me a bit nervous; I would rather be more defensive here and avoid passing credentials around too widely.

We already do this on the server - any package you install could use the server-side hook to get the user password AFAIK.

**Comment by stubailo** [comment]
On the other hand, we can be conservative for now and only pass the method arguments to the server-side callback - the client side callback will still be useful and you can read from `Meteor.user()` to get necessary data.

**Comment by glasser** [comment]
I dunno, without the arguments you can't know (in `onLoginFailure`) what kind of login was attempted, for sites allowing multiple OAuth providers, or both OAuth and password.

**Comment by jakozaur** [comment]
Would second @stubailo proposal: We don't pass attempt argument on client. Added commit and reran the tests.

My thinking:
1. "When in doubt leave it out." (common API wisdom)
2. Keep it simple.
3. Don't see any valid use case why we need attempt object on client. Yes, theoretically we may want to know what kind of login was attempted, but lets wait for real need. @glasser have you experienced application when you take advantage of that information on client? 
4. Though the original code is correct secure wise, the framework should also promote right usage patterns. @justinsb is right. Passing passwords around makes it too easy to shoot yourself in the foot.
5. I love Meteor. Thank you for your input!

**Comment by jakozaur** [comment]
Ping @stubailo :-). Let me know if there is anything I can do to get this request merged.

**Comment by stubailo** [comment]
We're about to have an off-focus week, I'll try to make sure this PR gets merged then. Sorry for the delay!

**Comment by stubailo** [comment]
Merged: https://github.com/meteor/meteor/commit/b5861c114ab2b8f76c36d9e4ed2ada84903732e4

We can add the arguments back in later if we need to.

**Comment by jakozaur** [comment]
Thanks for merging @stubailo !

Happy

**Pull request closed without merging**